File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode newhead;
    public ListNode reverse(ListNode node) {
        if(node.next == null) {
            newhead = node;
            return node;
        }
        ListNode prev = reverse(node.next);
        prev.next = node;
        node.next = null;
        return node;
    }
    public ListNode reverseList(ListNode head) {
        if(head == null || head.next == null) {
            return head;
        }
        reverse(head);
        return newhead;
        
    }
}

File: 0206-reverse-linked-list/0206-reverse-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    
    let newhead = null
    let dfs = (p) => {
        if(p) {
            let prev = dfs(p.next)
            //console.log(next ? next.val : null)
            if(prev == null) {
                newhead = p
            } else {
                p.next = null
                prev.next = p
            }
            return p
        }
    }
    
    dfs(head)
    return newhead
};